benchmark/benchmark.js

```diff
@@ -0,0 +1,453 @@
+ *
+ * @param {string} uri
+ * @returns {Promise<FileInfo>}
+ *
+ * @typedef {{ uri: string, path: string, size: number }} FileInfo
+ */
+function getFileInfo(uri) {
+	let info = fileInfoCache.get(uri);
+	if (info === undefined) {
+		info = getFileInfoUncached(uri);
+		fileInfoCache.set(uri, info);
+	}
+	return info;
+}
+/**
+ * @param {string} uri
+ * @returns {Promise<FileInfo>}
+ */
+async function getFileInfoUncached(uri) {
+	const p = await getFilePath(uri);
+	const stat = await fs.promises.stat(p);
+	if (stat.isFile()) {
+		return {
+			uri,
+			path: p,
+			size: stat.size
+		};
+	} else {
+		throw new Error(`Unknown file "${uri}"`);
+	}
+}
+/**
+ * Returns the local path of the given file identifier.
+ *
+ * @param {string} uri
+ * @returns {Promise<string>}
+ */
+async function getFilePath(uri) {
+	if (/^https:\/\//.test(uri)) {
+		// it's a URL, so let's download the file (if not downloaded already)
+		const downloadDir = path.join(__dirname, 'downloads');
+		await fs.promises.mkdir(downloadDir, { recursive: true });
+
+		// file path
+		const hash = crypto.createHash('md5').update(uri).digest('hex');
+		const localPath = path.resolve(downloadDir, hash + '-' + /[-\w\.]*$/.exec(uri)[0]);
+
+		if (!fs.existsSync(localPath)) {
+			// download file
+			console.log(`Downloading ${uri}...`);
+			await fs.promises.writeFile(localPath, await fetch(uri).then(r => r.text()), 'utf8');
+		}
+
+		return localPath;
+	}
+
+	// assume that it's a local file
+	return path.resolve(__dirname, uri);
+}
+
+/**
+ * @param {Iterable<[string, () => void]>} candidates
+ * @param {import("benchmark").Options} [options]
+ * @returns {Result[]}
+ *
+ * @typedef {{ name: string, stats: import("benchmark").Stats }} Result
+ */
+function measureCandidates(candidates, options) {
+	const suite = new Benchmark.Suite('temp name');
+
+	for (const [name, fn] of candidates) {
+		suite.add(name, fn, options);
+	}
+
+	/** @type {Result[]} */
+	const results = [];
+
+	suite.on('cycle', event => {
+		results.push({
+			name: event.target.name,
+			stats: event.target.stats
+		});
+	}).run();
+
+	return results;
+}
+
+/**
+ * @param {Result[]} results
+ * @returns {Result | null}
+ */
+function getBest(results) {
+	if (results.length >= 2) {
+		const sorted = [...results].sort((a, b) => a.stats.mean - b.stats.mean);
+		const best = sorted[0].stats;
+		const secondBest = sorted[1].stats;
+
+		// basically, it's only the best if the two means plus their moe are disjoint
+		if (best.mean + best.moe + secondBest.moe < secondBest.mean) {
+			return sorted[0];
+		}
+	}
+
+	return null;
+}
+/**
+ * @param {Result[]} results
+ * @returns {Result | null}
+ */
+function getWorst(results) {
+	if (results.length >= 2) {
+		const sorted = [...results].sort((a, b) => b.stats.mean - a.stats.mean);
+		const worst = sorted[0].stats;
+		const secondWorst = sorted[1].stats;
+
+		// basically, it's only the best if the two means plus their moe are disjoint
```

Comment:
ngl i have no idea what this means. what's a "moe"?

Reply:
No idea. It's been like a year. I'll try to fix it.

I only know that it's not what Google thinks it is lol.

Reply:
Now I remember. moe = [margin of error](https://benchmarkjs.com/docs#stats_moe).